build-tools/packages/build-cli/src/library/markdown.ts

```diff
@@ -26,18 +27,63 @@ export function remarkHeadingLinks(): (tree: Node) => void {
 			if (node.children?.length > 0) {
 				const firstChild = node.children[0];
 				if (firstChild.type === "text") {
-					const headingValue = firstChild.value;
-					const slug = slugger.slug(headingValue);
+					const slug = slugger.slug(toString(node));
 					// We need to insert an Html node instead of a string, because raw
 					// strings will get markdown-escaped when rendered
 					const htmlNode: Html = {
 						type: "html",
 						value: `<a id="${slug}"></a>`,
 					};
-					// Insert the HTML node before the text node of the heading
+					// Insert the HTML node as the first child node of the heading
 					node.children.unshift(htmlNode);
 				}
 			}
 		});
 	};
 }
+
+/**
+ * A regular expression that extracts an admonition title from a string UNLESS the admonition title is the only thing on
+ * the line.
+ *
+ * Capture group 1 is the admonition type/title. Capture group 2 is any trailing whitespace.
+ *
+ * @remarks
+ *
+ * Description of the regular expression:
+ *
+ * This regular expression matches patterns in the form of `[!WORD]` where WORD can be CAUTION, IMPORTANT, NOTE, TIP, or
+ * WARNING. It ensures that the pattern is not followed by only whitespace characters until the end of the line.
+ * Additionally, it captures any whitespace characters that follow the matched pattern.
+ */
+const ADMONITION_REGEX = /(\[!(?:CAUTION|IMPORTANT|NOTE|TIP|WARNING)])(?!\s*$)(\s*)/gm;
+
+/**
+ * A regular expression to remove single line breaks from text. This is used to remove extraneous line breaks in text
+ * nodes in markdown. This is useful because GitHub sometimes renders single line breaks, and sometimes it ignores them
+ * like the CommonMark spec describes. Removing them ensures that markdown renders as expected across GitHub.
+ */
+const SOFT_BREAK_REGEX = /$[^$]/gms;
```

Comment:
I tried it in regex101.com and it works but I don't quite understand _why_. Look for an end-of-line (because in multiline mode, so not just end-of-string)... where the next character is not _another_ (?) end-of-line? Or in this case interpreted as end-of-string, so "an end-of-line where the next character is not the end of the string"?

Reply:
> an end-of-line where the next character is not the end of the string

Yeah, but it's a bit sneaky. The $ in the `[^$]` piece could be replaced with almost any character (`&` for example), because it's interpreted literally in the brackets. So the regex essentially finds the end of lines then captures another single character that isn't the literal `$` - which would be the newline itself. I found this on another GitHub releases tool, so I know it's battle-tested, but `$\s*` is maybe equivalent and easy to understand?

Reply:
Maybe. On an actual end-of-file with trailing whitespace I think it's still fine to match and replace it, right?

But if it's in another tool, maybe just commenting "This is tricky to understand but battle-tested as proven by <link to tool>" would be fine as well.

Reply:
Added a comment and a link.